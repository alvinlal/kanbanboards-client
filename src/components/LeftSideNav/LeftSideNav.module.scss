.left__side__nav {
  width: 240px;
  height: calc(100vh - 70px);
  padding-top: 40px;
  background-color: $grey-50;
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  .heading {
    margin-bottom: 16px;

    h6 {
      color: $text-300;
      font-weight: $bold;
    }

    svg {
      color: $text-300;
      transform: rotate(45deg);
      display: none;
      cursor: pointer;
    }
    padding: 0px 24px;
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  .error {
    display: flex;
    align-items: center;
    justify-content: center;
    color: $error-300;
    width: 100%;

    p {
      margin-left: 5px;
    }
  }

  .no__boards {
    margin-left: 25px;
    margin-top: 40px;
  }

  .logout__btn {
    display: none;
    margin-bottom: 10px;
  }

  @media screen and (max-width: $sm) {
    height: 100vh;
    max-height: 100vh;
    position: fixed;
    top: 0px;
    position: absolute;
    width: 100%;
    left: -100%;
    transition: left 0.2s linear;

    .heading {
      margin-top: 40px;
      svg {
        display: block;
      }
    }

    .logout__btn {
      display: flex;
    }
  }
}
