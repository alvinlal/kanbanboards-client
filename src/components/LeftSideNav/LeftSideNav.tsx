/* eslint-disable no-nested-ternary */
import {
  ArrowRightOnRectangleIcon,
  ExclamationTriangleIcon,
  PlusIcon,
} from '@heroicons/react/24/outline';
import { useMemo } from 'react';
import Button from '../Button/Button';
import SkeletonBlock from '../SkeletonBlock/SkeletonBlock';
import BoardList from './components/BoardList/BoardList';
import { useAllBoards } from './hooks/useAllBoards';
import { useLeftSideNav } from './hooks/useLeftSideNav';
import styles from './LeftSideNav.module.scss';

const LeftSideNav: React.FC = () => {
  const { isLeftSideNavOpen, toggleLeftSideNav, logOut } = useLeftSideNav();
  const { allBoards, isLoading, isError } = useAllBoards();

  const LoadingBlocks = useMemo(() => {
    const blocks = [];
    for (let i = 0; i < 6; i += 1) {
      blocks.push(<SkeletonBlock width="200px" height="30px" key={i} />);
    }
    return blocks;
  }, []);

  return (
    <div className={styles.left__side__nav} style={{ left: isLeftSideNavOpen ? 0 : '-100%' }}>
      <div>
        <div className={styles.heading}>
          <h6>All Boards</h6>
          <PlusIcon width={32} height={32} strokeWidth={2} onClick={toggleLeftSideNav} />
        </div>
        {isLoading && <div data-testid="skeleton-blocks"> {LoadingBlocks}</div>}
        {isError && (
          <div className={styles.error}>
            <ExclamationTriangleIcon width={28} height={28} />
            <p>Something went wrong !</p>
          </div>
        )}
        {allBoards ? (
          allBoards.length ? (
            <BoardList allBoards={allBoards} />
          ) : (
            <p className={styles.no__boards}>No boards found</p>
          )
        ) : null}
      </div>
      <div>
        <div className={styles.logout__btn}>
          <Button
            style={{ height: '64px', width: '100%', justifyContent: 'space-between' }}
            onClick={logOut}
          >
            <h6>Logout</h6>
            <ArrowRightOnRectangleIcon width={32} height={32} strokeWidth={2} />
          </Button>
        </div>
        <Button
          style={{ height: '64px', width: '100%', justifyContent: 'space-between', zIndex: 10 }}
        >
          <h6>New Board</h6>
          <PlusIcon width={32} height={32} strokeWidth={2} />
        </Button>
      </div>
    </div>
  );
};

export default LeftSideNav;
