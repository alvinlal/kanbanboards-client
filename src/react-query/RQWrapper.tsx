/* eslint-disable no-console */
// this wrapper is for testing only

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import React from 'react';

const queryClient = new QueryClient({
  logger: {
    log: console.log,
    warn: console.warn,
    error: () => {},
  },
});
export const RQWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
