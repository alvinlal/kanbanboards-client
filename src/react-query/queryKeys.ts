const queryKeys = {
  user: 'USER',
};

export default queryKeys;
