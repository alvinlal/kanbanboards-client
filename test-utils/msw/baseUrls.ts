/* eslint-disable import/prefer-default-export */

export const apiEndPoint = (path: string) => {
  return import.meta.env.VITE_API_ENDPOINT + path;
};
