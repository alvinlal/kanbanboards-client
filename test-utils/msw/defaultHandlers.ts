import { ResponseComposition, RestContext, RestRequest } from 'msw';

export default {
  LOADING: (_: RestRequest, res: ResponseComposition, ctx: RestContext) =>
    res(ctx.delay('infinite')),
  500: (_: RestRequest, res: ResponseComposition, ctx: RestContext) =>
    res(
      ctx.status(500),
      ctx.json({
        statusCode: 500,
        message: 'Internal server error',
      })
    ),
  ERR_NETWORK: (_: RestRequest, res: ResponseComposition) =>
    res.networkError('failed to connect'),
};
