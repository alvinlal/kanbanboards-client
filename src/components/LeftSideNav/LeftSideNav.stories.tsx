import { ComponentStory, ComponentMeta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import { rest } from 'msw';
import { apiEndPoint } from '../../../test-utils/msw/baseUrls';
import defaultHandlers from '../../../test-utils/msw/defaultHandlers';
import LeftSideNavigation from './LeftSideNav';

export default {
  title: 'Components/LeftSideNav',
  component: LeftSideNavigation,
  decorators: [
    (Story) => (
      <div
        style={{
          width: '100%',
          height: '100vh',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof LeftSideNavigation>;

export const Default: ComponentStory<typeof LeftSideNavigation> = () => (
  <MemoryRouter>
    <LeftSideNavigation />
  </MemoryRouter>
);

export const Loading: ComponentStory<typeof LeftSideNavigation> = () => (
  <MemoryRouter>
    <LeftSideNavigation />
  </MemoryRouter>
);

Loading.parameters = {
  msw: {
    handlers: [rest.get(apiEndPoint('/boards/all'), defaultHandlers.LOADING)],
  },
};

export const Empty: ComponentStory<typeof LeftSideNavigation> = () => (
  <MemoryRouter>
    <LeftSideNavigation />
  </MemoryRouter>
);

Empty.parameters = {
  msw: {
    handlers: [
      rest.get(apiEndPoint('/boards/all'), (_, res, ctx) => {
        return res(ctx.json([]));
      }),
    ],
  },
};

export const Error: ComponentStory<typeof LeftSideNavigation> = () => (
  <MemoryRouter>
    <LeftSideNavigation />
  </MemoryRouter>
);

Error.parameters = {
  msw: {
    handlers: [rest.get(apiEndPoint('/boards/all'), defaultHandlers[500])],
  },
};
