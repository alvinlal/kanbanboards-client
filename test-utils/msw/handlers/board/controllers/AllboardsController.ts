import { RestRequest, ResponseComposition, RestContext } from 'msw';
import { allBoardsResponse } from '../../../../stubs/allBoardsResponse.stub';
import defaultHandlers from '../../../defaultHandlers';

export const AllboardsController = {
  ...defaultHandlers,
  200: (_: RestRequest, res: ResponseComposition, ctx: RestContext) =>
    res(ctx.json(allBoardsResponse)),
  EMPTY: (_: RestRequest, res: ResponseComposition, ctx: RestContext) => res(ctx.json([])),
};
