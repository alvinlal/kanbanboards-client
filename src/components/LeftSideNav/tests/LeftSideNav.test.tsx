/* eslint-disable no-console */
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { apiEndPoint } from '../../../../test-utils/msw/baseUrls';
import defaultHandlers from '../../../../test-utils/msw/defaultHandlers';
import { AllboardsController } from '../../../../test-utils/msw/handlers/board/controllers/AllboardsController';
import { rest, server } from '../../../../test-utils/msw/server';
import LeftSideNav from '../LeftSideNav';

describe('LeftSideNav.tsx', () => {
  const queryClient = new QueryClient({
    logger: {
      log: console.log,
      warn: console.warn,
      error: () => {},
    },
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  beforeEach(() => {
    queryClient.clear();
  });
  it('should show skeleton blocks while loading', async () => {
    server.use(rest.get(apiEndPoint('/boards/all'), defaultHandlers.LOADING));

    render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LeftSideNav />
        </MemoryRouter>
      </QueryClientProvider>
    );

    const skeletonBlocks = await screen.findByTestId('skeleton-blocks');

    expect(skeletonBlocks).toBeInTheDocument();
  });

  it('should show error message on error', async () => {
    server.use(rest.get(apiEndPoint('/boards/all'), defaultHandlers[500]));

    render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LeftSideNav />
        </MemoryRouter>
      </QueryClientProvider>
    );

    expect(await screen.findByText('Something went wrong !')).toBeInTheDocument();
  });

  it('should show correct message when board list is empty', async () => {
    server.use(rest.get(apiEndPoint('/boards/all'), AllboardsController.EMPTY));

    render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LeftSideNav />
        </MemoryRouter>
      </QueryClientProvider>
    );

    expect(await screen.findByText('No boards found')).toBeInTheDocument();
  });

  it('should match snapshot in loading state', () => {
    server.use(rest.get(apiEndPoint('/boards/all'), defaultHandlers.LOADING));

    const tree = renderer.create(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LeftSideNav />
        </MemoryRouter>
      </QueryClientProvider>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot in error state', async () => {
    server.use(rest.get(apiEndPoint('/boards/all'), AllboardsController[500]));

    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LeftSideNav />
        </MemoryRouter>
      </QueryClientProvider>
    );

    await waitFor(() => {
      expect(screen.getByText('Something went wrong !')).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it.todo('should add a new board');

  it.todo('should delete a board');
});
