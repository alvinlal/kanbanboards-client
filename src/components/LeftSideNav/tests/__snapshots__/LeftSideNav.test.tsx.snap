// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftSideNav.tsx should match snapshot in error state 1`] = `
<DocumentFragment>
  <div
    class="left__side__nav"
    style="left: -100%;"
  >
    <div>
      <div
        class="heading"
      >
        <h6>
          All Boards
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.5v15m7.5-7.5h-15"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      <div
        class="error"
      >
        <svg
          aria-hidden="true"
          fill="none"
          height="28"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 9v3.75m-9.303 3.376c-.866 1.5.217 3.374 1.948 3.374h14.71c1.73 0 2.813-1.874 1.948-3.374L13.949 3.378c-.866-1.5-3.032-1.5-3.898 0L2.697 16.126zM12 15.75h.007v.008H12v-.008z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <p>
          Something went wrong !
        </p>
      </div>
    </div>
    <div>
      <div
        class="logout__btn"
      >
        <button
          class="button primary "
          style="height: 64px; width: 100%; justify-content: space-between;"
        >
          <h6>
            Logout
          </h6>
          <svg
            aria-hidden="true"
            fill="none"
            height="32"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.75 9V5.25A2.25 2.25 0 0013.5 3h-6a2.25 2.25 0 00-2.25 2.25v13.5A2.25 2.25 0 007.5 21h6a2.25 2.25 0 002.25-2.25V15m3 0l3-3m0 0l-3-3m3 3H9"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <button
        class="button primary "
        style="height: 64px; width: 100%; justify-content: space-between; z-index: 10;"
      >
        <h6>
          New Board
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.5v15m7.5-7.5h-15"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LeftSideNav.tsx should match snapshot in loading state 1`] = `
<div
  className="left__side__nav"
  style={
    {
      "left": "-100%",
    }
  }
>
  <div>
    <div
      className="heading"
    >
      <h6>
        All Boards
      </h6>
      <svg
        aria-hidden="true"
        fill="none"
        height={32}
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 4.5v15m7.5-7.5h-15"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    </div>
    <div
      data-testid="skeleton-blocks"
    >
       
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
      <div
        className="skeleton__block"
        style={
          {
            "height": "30px",
            "width": "200px",
          }
        }
      />
    </div>
  </div>
  <div>
    <div
      className="logout__btn"
    >
      <button
        className="button primary "
        onClick={[Function]}
        style={
          {
            "height": "64px",
            "justifyContent": "space-between",
            "width": "100%",
          }
        }
      >
        <h6>
          Logout
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height={32}
          stroke="currentColor"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.75 9V5.25A2.25 2.25 0 0013.5 3h-6a2.25 2.25 0 00-2.25 2.25v13.5A2.25 2.25 0 007.5 21h6a2.25 2.25 0 002.25-2.25V15m3 0l3-3m0 0l-3-3m3 3H9"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
    </div>
    <button
      className="button primary "
      style={
        {
          "height": "64px",
          "justifyContent": "space-between",
          "width": "100%",
          "zIndex": 10,
        }
      }
    >
      <h6>
        New Board
      </h6>
      <svg
        aria-hidden="true"
        fill="none"
        height={32}
        stroke="currentColor"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 4.5v15m7.5-7.5h-15"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;
